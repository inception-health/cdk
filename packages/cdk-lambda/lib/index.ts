export { Lambda } from "./lambda";
export type { LambdaProps } from "./lambda";

export { DeadLetterQueue } from "./dlq/queue";
export type { DeadLetterQueueProps } from "./dlq/queue";

export { StandardNetwork } from "./network/network";
export type { StandardNetworkProps } from "./network/network";
