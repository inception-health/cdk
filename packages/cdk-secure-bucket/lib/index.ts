export * from "./s3.secure";
export * from "./lifecycle";
