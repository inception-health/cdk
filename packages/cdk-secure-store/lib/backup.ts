import { aws_backup, aws_dynamodb, aws_kms } from "aws-cdk-lib";
import { Construct } from "constructs";

export interface SecureStoreBackupProps {
  backupPlanId?: string;
  encryptionKey: aws_kms.IKey;
  table: aws_dynamodb.ITableV2;
  tableName: string;
}

export interface SecureStoreBackupResources {
  backupPlan: aws_backup.IBackupPlan | undefined;
  backupVault: aws_backup.BackupVault | undefined;
}

export class SecureStoreBackup extends Construct {
  public readonly cdk: SecureStoreBackupResources;

  private readonly props: SecureStoreBackupProps;

  constructor(scope: Construct, id: string, props: SecureStoreBackupProps) {
    super(scope, id);

    // Initialize the properties
    this.props = props;
    this.cdk = {} as SecureStoreBackupResources;

    // Create the backup resources
    if (this.props.backupPlanId) {
      this.cdk.backupPlan = aws_backup.BackupPlan.fromBackupPlanId(
        this,
        "backupPlan",
        this.props.backupPlanId,
      );
    } else {
      // Create a backup vault
      this.cdk.backupVault = new aws_backup.BackupVault(this, "backupVault", {
        backupVaultName: `secure-store-${this.props.tableName}-backup-vault`,
        encryptionKey: this.props.encryptionKey,
        blockRecoveryPointDeletion: true,
        // TODO: Review if we want to make lock configuration by default.
        // lockConfiguration
      });

      // Create the backup plan
      this.cdk.backupPlan =
        aws_backup.BackupPlan.dailyWeeklyMonthly7YearRetention(
          this,
          `secure-store-${this.props.tableName}-backup-plan`,
          this.cdk.backupVault,
        );

      (this.cdk.backupPlan as aws_backup.BackupPlan).addSelection(
        `secure-store-${this.props.tableName}-backup-plan-selection`,
        {
          resources: [
            aws_backup.BackupResource.fromDynamoDbTable(this.props.table),
          ],
        },
      );
    }
  }
}
