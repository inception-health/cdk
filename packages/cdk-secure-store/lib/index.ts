export type { SecureStoreProps } from "./store";
export { SecureStore } from "./store";
