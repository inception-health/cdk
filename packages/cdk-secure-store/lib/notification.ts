import {
  aws_events,
  aws_events_targets,
  aws_kms,
  aws_logs,
  aws_sns,
} from "aws-cdk-lib";
import { Construct } from "constructs";

export interface SecureStoreNotificationProps {
  /**
   * The SNS Topic to send notifications to.
   */
  destination: string;
  /**
   * Enable CloudWatch Logs for the EventBridge rule.
   *
   * @defaultValue false
   */
  enableLogs?: boolean;
  /**
   * Enable or disable the notification.
   *
   * @defaultValue true
   */
  enabled?: boolean;
  /**
   * The encryption key to use for encrypting the sns topic and the logs
   *
   * @remarks
   * The encryption key must be created in the same account and region as the
   * DynamoDB table. It also must have the following policy attached to it:
   * ```
   * {
   *   "Sid": "Allow access to CloudWatch Log",
   *   "Effect": "Allow",
   *   "Principal": {
   *       "Service": "logs.REGION.amazonaws.com"
   *   },
   *   "Action": [
   *       "kms:Encrypt",
   *       "kms:Decrypt",
   *       "kms:ReEncrypt*",
   *       "kms:GenerateDataKey*",
   *       "kms:DescribeKey"
   *   ],
   *   "Resource": "*",
   *   "Condition": {
   *       "ArnLike": {
   *           "kms:EncryptionContext:aws:logs:arn": "arn:aws:logs:REGION:ACCOUNT_ID:*"
   *       }
   *   }
   * }
   *```
   */
  encryptionKey: aws_kms.IKey;
  /**
   * The secure store table name
   */
  sourceTable: string;
}

const defaultProps: Partial<SecureStoreNotificationProps> = {
  enabled: true,
  enableLogs: true,
};

interface SecureStoreNotificationResources {
  bus: aws_events.IEventBus;
  logs: aws_logs.LogGroup;
  rule: aws_events.Rule;
  topic: aws_sns.ITopic;
}

/**
 * `Notification` is a construct that creates a notification for specific
 *  DynamoDB events. The notification is sent to the provided SNS topic.
 *
 * At the moment, the notification is sent for the following events:
 * - CreateTable
 * - DeleteTable
 * - UpdateTable
 * - DeleteBackup
 * - RestoreTableFromBackup
 * - RestoreTableToPointInTime
 *
 * @remarks
 * For the full list of events that trigger a notification, see the [DynamoDB
 * documentation](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/logging-using-cloudtrail.html).
 */
export class SecureStoreNotification extends Construct {
  public readonly cdk: SecureStoreNotificationResources;

  // !Private Properties
  private readonly props: SecureStoreNotificationProps;

  // !Constructor
  constructor(
    scope: Construct,
    id: string,
    props: SecureStoreNotificationProps,
  ) {
    super(scope, id);

    // Initialize the properties
    this.props = { ...defaultProps, ...props };
    this.cdk = {} as SecureStoreNotificationResources;

    // Create Resources
    this.setupTopic();

    this.setupBus();

    this.setupRule();

    this.setupLogs();
  }

  // !Private Methods
  private setupTopic() {
    this.cdk.topic = aws_sns.Topic.fromTopicArn(
      this,
      "topic",
      this.props.destination,
    );
  }

  private setupBus() {
    this.cdk.bus = aws_events.EventBus.fromEventBusName(
      this,
      "eventBus",
      "default",
    );
  }

  private setupRule() {
    this.cdk.rule = new aws_events.Rule(this, "rule", {
      ruleName: `${this.props.sourceTable}-dynamodb--monitoring`,
      description: `Send system notifications related to the ${this.props.sourceTable} table.`,
      enabled: this.props.enabled,
      eventBus: this.cdk.bus,
      eventPattern: {
        source: ["aws.dynamodb"],
        detailType: ["AWS API Call via CloudTrail"],
        detail: {
          eventSource: ["dynamodb.amazonaws.com"],
          eventName: [
            "CreateTable",
            "DeleteTable",
            "UpdateTable",
            "DeleteBackup",
            "RestoreTableFromBackup",
            "RestoreTableToPointInTime",
          ],
          requestParameters: {
            tableName: [this.props.sourceTable],
          },
        },
      },
    });

    // Add the target to the rule
    this.cdk.rule.addTarget(new aws_events_targets.SnsTopic(this.cdk.topic));
  }

  private setupLogs() {
    if (!this.props.enableLogs) return;

    // New Cloudwatch Log Group to deliver EventBridge Events to.
    this.cdk.logs = new aws_logs.LogGroup(this, "logs", {
      logGroupName: `/inception/devops/${this.props.sourceTable}-dynamodb--monitoring`,
      retention: aws_logs.RetentionDays.TWO_WEEKS,
      encryptionKey: this.props.encryptionKey,
    });

    // Define rule that will deliver EventBridge events to a Cloudwatch Log Group
    this.cdk.rule.addTarget(
      new aws_events_targets.CloudWatchLogGroup(this.cdk.logs),
    );
  }
}
