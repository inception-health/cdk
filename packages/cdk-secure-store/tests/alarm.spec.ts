import { aws_kms, aws_sns } from "aws-cdk-lib";
import { Match, Template } from "aws-cdk-lib/assertions";
import { Stack } from "aws-cdk-lib/core";

import { SecureStoreAlarms } from "../lib/alarm";
import { SecureStore } from "../lib/store";

describe("SecureStoreAlarms", () => {
  let stack: Stack;
  let store: SecureStore;
  let topic: aws_sns.ITopic;

  beforeEach(() => {
    stack = new Stack(undefined, undefined, {
      env: { region: "us-east-1" },
    });
    store = new SecureStore(stack, "store", {
      encryptionKey: new aws_kms.Key(stack, "key"),
      tableName: "table",
    });
    topic = aws_sns.Topic.fromTopicArn(
      stack,
      "topic",
      "arn:aws:sns:us-east-1:123456789012:topic",
    );
  });

  it("should match the snapshot", () => {
    new SecureStoreAlarms(store, "MySecureStoreAlarms", {
      alertsTopic: topic.topicArn,
      systemErrorAlarm: true,
      throttleAlarm: true,
    });

    // then
    expect(Template.fromStack(stack)).toMatchSnapshot();
  });

  it("should create a new instance of SecureStoreAlarms", () => {
    const secureStoreAlarms = new SecureStoreAlarms(
      store,
      "MySecureStoreAlarms",
      {
        alertsTopic: topic.topicArn,
        systemErrorAlarm: true,
        throttleAlarm: true,
      },
    );

    expect(secureStoreAlarms).toBeInstanceOf(SecureStoreAlarms);
  });

  it("should setup system error alarm if systemErrorAlarm prop is true", () => {
    new SecureStoreAlarms(store, "MySecureStoreAlarms", {
      alertsTopic: topic.topicArn,
      systemErrorAlarm: true,
      throttleAlarm: false,
    });

    // Then
    // Template.fromStack(stack).resourceCountIs("AWS::CloudWatch::Alarm", 1);
    Template.fromStack(stack).hasResourceProperties("AWS::CloudWatch::Alarm", {
      ComparisonOperator: "GreaterThanOrEqualToThreshold",
      DatapointsToAlarm: 1,
      EvaluationPeriods: 6,
      Metrics: Match.arrayWith([
        {
          Id: "getitem",
          MetricStat: {
            Metric: Match.objectLike({
              MetricName: "SystemErrors",
              Namespace: "AWS/DynamoDB",
            }),
            Period: 300,
            Stat: "Sum",
          },
          ReturnData: false,
        },
      ]),
    });
  });

  it("should setup throttle alarm if throttleAlarm prop is true", () => {
    new SecureStoreAlarms(store, "MySecureStoreAlarms", {
      alertsTopic: topic.topicArn,
      systemErrorAlarm: false,
      throttleAlarm: true,
    });

    // Then
    // Template.fromStack(stack).resourceCountIs("AWS::CloudWatch::Alarm", 1);
    Template.fromStack(stack).hasResourceProperties("AWS::CloudWatch::Alarm", {
      ComparisonOperator: "GreaterThanOrEqualToThreshold",
      DatapointsToAlarm: 1,
      EvaluationPeriods: 6,
      Metrics: Match.arrayWith([
        {
          Id: "getitem",
          MetricStat: {
            Metric: Match.objectLike({
              MetricName: "SystemErrors",
              Namespace: "AWS/DynamoDB",
            }),
            Period: 300,
            Stat: "Sum",
          },
          ReturnData: false,
        },
      ]),
    });
  });
});
