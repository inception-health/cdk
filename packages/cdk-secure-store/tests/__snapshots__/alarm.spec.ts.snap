// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SecureStoreAlarms should match the snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "keyFEDD6EC0": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "KeyPolicy": {
          "Statement": [
            {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::Join": [
                    "",
                    [
                      "arn:",
                      {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "storeMySecureStoreAlarmsalarmerrorsE0E0F65D": {
      "Properties": {
        "AlarmActions": [
          "arn:aws:sns:us-east-1:123456789012:topic",
        ],
        "AlarmName": {
          "Fn::Join": [
            "",
            [
              "DynamoDB ",
              {
                "Ref": "storetableE8648AF5",
              },
              " System Errors",
            ],
          ],
        },
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "DatapointsToAlarm": 1,
        "EvaluationPeriods": 6,
        "Metrics": [
          {
            "Expression": "getitem + putitem + deleteitem + updateitem + transactwriteitems + transactgetitems + batchgetitem + batchwriteitem + query + scan",
            "Id": "expr_1",
            "Label": "Sum of errors across all operations",
          },
          {
            "Id": "getitem",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "Operation",
                    "Value": "GetItem",
                  },
                  {
                    "Name": "TableName",
                    "Value": {
                      "Ref": "storetableE8648AF5",
                    },
                  },
                ],
                "MetricName": "SystemErrors",
                "Namespace": "AWS/DynamoDB",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
          {
            "Id": "putitem",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "Operation",
                    "Value": "PutItem",
                  },
                  {
                    "Name": "TableName",
                    "Value": {
                      "Ref": "storetableE8648AF5",
                    },
                  },
                ],
                "MetricName": "SystemErrors",
                "Namespace": "AWS/DynamoDB",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
          {
            "Id": "deleteitem",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "Operation",
                    "Value": "DeleteItem",
                  },
                  {
                    "Name": "TableName",
                    "Value": {
                      "Ref": "storetableE8648AF5",
                    },
                  },
                ],
                "MetricName": "SystemErrors",
                "Namespace": "AWS/DynamoDB",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
          {
            "Id": "updateitem",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "Operation",
                    "Value": "UpdateItem",
                  },
                  {
                    "Name": "TableName",
                    "Value": {
                      "Ref": "storetableE8648AF5",
                    },
                  },
                ],
                "MetricName": "SystemErrors",
                "Namespace": "AWS/DynamoDB",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
          {
            "Id": "transactwriteitems",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "Operation",
                    "Value": "TransactWriteItems",
                  },
                  {
                    "Name": "TableName",
                    "Value": {
                      "Ref": "storetableE8648AF5",
                    },
                  },
                ],
                "MetricName": "SystemErrors",
                "Namespace": "AWS/DynamoDB",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
          {
            "Id": "transactgetitems",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "Operation",
                    "Value": "TransactGetItems",
                  },
                  {
                    "Name": "TableName",
                    "Value": {
                      "Ref": "storetableE8648AF5",
                    },
                  },
                ],
                "MetricName": "SystemErrors",
                "Namespace": "AWS/DynamoDB",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
          {
            "Id": "batchgetitem",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "Operation",
                    "Value": "BatchGetItem",
                  },
                  {
                    "Name": "TableName",
                    "Value": {
                      "Ref": "storetableE8648AF5",
                    },
                  },
                ],
                "MetricName": "SystemErrors",
                "Namespace": "AWS/DynamoDB",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
          {
            "Id": "batchwriteitem",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "Operation",
                    "Value": "BatchWriteItem",
                  },
                  {
                    "Name": "TableName",
                    "Value": {
                      "Ref": "storetableE8648AF5",
                    },
                  },
                ],
                "MetricName": "SystemErrors",
                "Namespace": "AWS/DynamoDB",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
          {
            "Id": "query",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "Operation",
                    "Value": "Query",
                  },
                  {
                    "Name": "TableName",
                    "Value": {
                      "Ref": "storetableE8648AF5",
                    },
                  },
                ],
                "MetricName": "SystemErrors",
                "Namespace": "AWS/DynamoDB",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
          {
            "Id": "scan",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "Operation",
                    "Value": "Scan",
                  },
                  {
                    "Name": "TableName",
                    "Value": {
                      "Ref": "storetableE8648AF5",
                    },
                  },
                ],
                "MetricName": "SystemErrors",
                "Namespace": "AWS/DynamoDB",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
        ],
        "Threshold": 0,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "storeMySecureStoreAlarmsalarmthrottleAC82B9D3": {
      "Properties": {
        "AlarmActions": [
          "arn:aws:sns:us-east-1:123456789012:topic",
        ],
        "AlarmName": {
          "Fn::Join": [
            "",
            [
              "DynamoDB ",
              {
                "Ref": "storetableE8648AF5",
              },
              " Throttled",
            ],
          ],
        },
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "DatapointsToAlarm": 1,
        "EvaluationPeriods": 6,
        "Metrics": [
          {
            "Expression": "m1 + m2",
            "Id": "expr_1",
            "Label": {
              "Fn::Join": [
                "",
                [
                  "DynamoDB ",
                  {
                    "Ref": "storetableE8648AF5",
                  },
                  " Reads/Writes Throttled",
                ],
              ],
            },
          },
          {
            "Id": "m1",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "TableName",
                    "Value": {
                      "Ref": "storetableE8648AF5",
                    },
                  },
                ],
                "MetricName": "ReadThrottleEvents",
                "Namespace": "AWS/DynamoDB",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
          {
            "Id": "m2",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "TableName",
                    "Value": {
                      "Ref": "storetableE8648AF5",
                    },
                  },
                ],
                "MetricName": "WriteThrottleEvents",
                "Namespace": "AWS/DynamoDB",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
        ],
        "Threshold": 0,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "storealarmsalarmerrorsF5FB63CF": {
      "Properties": {
        "AlarmName": {
          "Fn::Join": [
            "",
            [
              "DynamoDB ",
              {
                "Ref": "storetableE8648AF5",
              },
              " System Errors",
            ],
          ],
        },
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "DatapointsToAlarm": 1,
        "EvaluationPeriods": 6,
        "Metrics": [
          {
            "Expression": "getitem + putitem + deleteitem + updateitem + transactwriteitems + transactgetitems + batchgetitem + batchwriteitem + query + scan",
            "Id": "expr_1",
            "Label": "Sum of errors across all operations",
          },
          {
            "Id": "getitem",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "Operation",
                    "Value": "GetItem",
                  },
                  {
                    "Name": "TableName",
                    "Value": {
                      "Ref": "storetableE8648AF5",
                    },
                  },
                ],
                "MetricName": "SystemErrors",
                "Namespace": "AWS/DynamoDB",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
          {
            "Id": "putitem",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "Operation",
                    "Value": "PutItem",
                  },
                  {
                    "Name": "TableName",
                    "Value": {
                      "Ref": "storetableE8648AF5",
                    },
                  },
                ],
                "MetricName": "SystemErrors",
                "Namespace": "AWS/DynamoDB",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
          {
            "Id": "deleteitem",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "Operation",
                    "Value": "DeleteItem",
                  },
                  {
                    "Name": "TableName",
                    "Value": {
                      "Ref": "storetableE8648AF5",
                    },
                  },
                ],
                "MetricName": "SystemErrors",
                "Namespace": "AWS/DynamoDB",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
          {
            "Id": "updateitem",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "Operation",
                    "Value": "UpdateItem",
                  },
                  {
                    "Name": "TableName",
                    "Value": {
                      "Ref": "storetableE8648AF5",
                    },
                  },
                ],
                "MetricName": "SystemErrors",
                "Namespace": "AWS/DynamoDB",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
          {
            "Id": "transactwriteitems",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "Operation",
                    "Value": "TransactWriteItems",
                  },
                  {
                    "Name": "TableName",
                    "Value": {
                      "Ref": "storetableE8648AF5",
                    },
                  },
                ],
                "MetricName": "SystemErrors",
                "Namespace": "AWS/DynamoDB",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
          {
            "Id": "transactgetitems",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "Operation",
                    "Value": "TransactGetItems",
                  },
                  {
                    "Name": "TableName",
                    "Value": {
                      "Ref": "storetableE8648AF5",
                    },
                  },
                ],
                "MetricName": "SystemErrors",
                "Namespace": "AWS/DynamoDB",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
          {
            "Id": "batchgetitem",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "Operation",
                    "Value": "BatchGetItem",
                  },
                  {
                    "Name": "TableName",
                    "Value": {
                      "Ref": "storetableE8648AF5",
                    },
                  },
                ],
                "MetricName": "SystemErrors",
                "Namespace": "AWS/DynamoDB",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
          {
            "Id": "batchwriteitem",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "Operation",
                    "Value": "BatchWriteItem",
                  },
                  {
                    "Name": "TableName",
                    "Value": {
                      "Ref": "storetableE8648AF5",
                    },
                  },
                ],
                "MetricName": "SystemErrors",
                "Namespace": "AWS/DynamoDB",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
          {
            "Id": "query",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "Operation",
                    "Value": "Query",
                  },
                  {
                    "Name": "TableName",
                    "Value": {
                      "Ref": "storetableE8648AF5",
                    },
                  },
                ],
                "MetricName": "SystemErrors",
                "Namespace": "AWS/DynamoDB",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
          {
            "Id": "scan",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "Operation",
                    "Value": "Scan",
                  },
                  {
                    "Name": "TableName",
                    "Value": {
                      "Ref": "storetableE8648AF5",
                    },
                  },
                ],
                "MetricName": "SystemErrors",
                "Namespace": "AWS/DynamoDB",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
        ],
        "Threshold": 0,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "storealarmsalarmthrottle1348B55B": {
      "Properties": {
        "AlarmName": {
          "Fn::Join": [
            "",
            [
              "DynamoDB ",
              {
                "Ref": "storetableE8648AF5",
              },
              " Throttled",
            ],
          ],
        },
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "DatapointsToAlarm": 1,
        "EvaluationPeriods": 6,
        "Metrics": [
          {
            "Expression": "m1 + m2",
            "Id": "expr_1",
            "Label": {
              "Fn::Join": [
                "",
                [
                  "DynamoDB ",
                  {
                    "Ref": "storetableE8648AF5",
                  },
                  " Reads/Writes Throttled",
                ],
              ],
            },
          },
          {
            "Id": "m1",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "TableName",
                    "Value": {
                      "Ref": "storetableE8648AF5",
                    },
                  },
                ],
                "MetricName": "ReadThrottleEvents",
                "Namespace": "AWS/DynamoDB",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
          {
            "Id": "m2",
            "MetricStat": {
              "Metric": {
                "Dimensions": [
                  {
                    "Name": "TableName",
                    "Value": {
                      "Ref": "storetableE8648AF5",
                    },
                  },
                ],
                "MetricName": "WriteThrottleEvents",
                "Namespace": "AWS/DynamoDB",
              },
              "Period": 300,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
        ],
        "Threshold": 0,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "storebackupbackupVaultF84FF6EB": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AccessPolicy": {
          "Statement": [
            {
              "Action": [
                "backup:DeleteRecoveryPoint",
                "backup:UpdateRecoveryPointLifecycle",
              ],
              "Effect": "Deny",
              "Principal": {
                "AWS": "*",
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "BackupVaultName": "secure-store-table-backup-vault",
        "EncryptionKeyArn": {
          "Fn::GetAtt": [
            "keyFEDD6EC0",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Backup::BackupVault",
      "UpdateReplacePolicy": "Retain",
    },
    "storebackupsecurestoretablebackupplanCD73E039": {
      "Properties": {
        "BackupPlan": {
          "BackupPlanName": "secure-store-table-backup-plan",
          "BackupPlanRule": [
            {
              "Lifecycle": {
                "DeleteAfterDays": 35,
              },
              "RuleName": "Daily",
              "ScheduleExpression": "cron(0 5 * * ? *)",
              "TargetBackupVault": {
                "Fn::GetAtt": [
                  "storebackupbackupVaultF84FF6EB",
                  "BackupVaultName",
                ],
              },
            },
            {
              "Lifecycle": {
                "DeleteAfterDays": 90,
              },
              "RuleName": "Weekly",
              "ScheduleExpression": "cron(0 5 ? * SAT *)",
              "TargetBackupVault": {
                "Fn::GetAtt": [
                  "storebackupbackupVaultF84FF6EB",
                  "BackupVaultName",
                ],
              },
            },
            {
              "Lifecycle": {
                "DeleteAfterDays": 2555,
                "MoveToColdStorageAfterDays": 90,
              },
              "RuleName": "Monthly7Year",
              "ScheduleExpression": "cron(0 5 1 * ? *)",
              "TargetBackupVault": {
                "Fn::GetAtt": [
                  "storebackupbackupVaultF84FF6EB",
                  "BackupVaultName",
                ],
              },
            },
          ],
        },
      },
      "Type": "AWS::Backup::BackupPlan",
    },
    "storebackupsecurestoretablebackupplansecurestoretablebackupplanselection4860B7E9": {
      "Properties": {
        "BackupPlanId": {
          "Fn::GetAtt": [
            "storebackupsecurestoretablebackupplanCD73E039",
            "BackupPlanId",
          ],
        },
        "BackupSelection": {
          "IamRoleArn": {
            "Fn::GetAtt": [
              "storebackupsecurestoretablebackupplansecurestoretablebackupplanselectionRole453B4CC7",
              "Arn",
            ],
          },
          "Resources": [
            {
              "Fn::GetAtt": [
                "storetableE8648AF5",
                "Arn",
              ],
            },
          ],
          "SelectionName": "secure-store-table-backup-plan-selection",
        },
      },
      "Type": "AWS::Backup::BackupSelection",
    },
    "storebackupsecurestoretablebackupplansecurestoretablebackupplanselectionRole453B4CC7": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "backup.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSBackupServiceRolePolicyForBackup",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "storetableE8648AF5": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "PK",
            "AttributeType": "S",
          },
          {
            "AttributeName": "SK",
            "AttributeType": "S",
          },
          {
            "AttributeName": "GSI1PK",
            "AttributeType": "S",
          },
          {
            "AttributeName": "GSI1SK",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "GlobalSecondaryIndexes": [
          {
            "IndexName": "GSI1",
            "KeySchema": [
              {
                "AttributeName": "GSI1PK",
                "KeyType": "HASH",
              },
              {
                "AttributeName": "GSI1SK",
                "KeyType": "RANGE",
              },
            ],
            "Projection": {
              "ProjectionType": "ALL",
            },
          },
        ],
        "KeySchema": [
          {
            "AttributeName": "PK",
            "KeyType": "HASH",
          },
          {
            "AttributeName": "SK",
            "KeyType": "RANGE",
          },
        ],
        "Replicas": [
          {
            "ContributorInsightsSpecification": {
              "Enabled": true,
            },
            "DeletionProtectionEnabled": true,
            "GlobalSecondaryIndexes": [
              {
                "ContributorInsightsSpecification": {
                  "Enabled": true,
                },
                "IndexName": "GSI1",
              },
            ],
            "PointInTimeRecoverySpecification": {
              "PointInTimeRecoveryEnabled": true,
            },
            "Region": "us-east-1",
            "SSESpecification": {
              "KMSMasterKeyId": {
                "Fn::GetAtt": [
                  "keyFEDD6EC0",
                  "Arn",
                ],
              },
            },
          },
        ],
        "SSESpecification": {
          "SSEEnabled": true,
          "SSEType": "KMS",
        },
        "TableName": "table",
      },
      "Type": "AWS::DynamoDB::GlobalTable",
      "UpdateReplacePolicy": "Retain",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
